internal/reflect/reflection.go

```diff
@@ -0,0 +1,141 @@
+package reflect
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// Into uses the data in `val` to populate `target`, using the reflection
+// package to recursively reflect into structs and slices. If `target` is an
+// AttributeValue, its assignment method will be used instead of reflecting. If
+// `target` is a tftypes.ValueConverter, the FromTerraformValue method will be
+// used instead of using reflection. Primitives are set using the val.As
+// method. Structs use reflection: each exported struct field must have a
+// "tfsdk" tag with the name of the field in the tftypes.Value, and all fields
+// in the tftypes.Value must have a corresponding property in the struct. Into
+// will be called for each struct field. Slices will have Into called for each
+// element.
+func Into(ctx context.Context, val tftypes.Value, target interface{}, opts Options) error {
+	v := reflect.ValueOf(target)
+	if v.Kind() != reflect.Ptr {
+		return fmt.Errorf("target must be a pointer, got %T, which is a %s", target, v.Kind())
+	}
+	result, err := buildReflectValue(ctx, val, v.Elem(), opts, tftypes.NewAttributePath())
+	if err != nil {
+		return err
+	}
+	v.Set(result)
+	return nil
+}
+
+// buildReflectValue constructs a reflect.Value of the same type as `target`,
+// populated with the data in `val`. It will defensively instantiate new values
+// to set, making it safe for use with pointer types which may be nil. It tries
+// to give consumers the ability to override its default behaviors wherever
+// possible.
+func buildReflectValue(ctx context.Context, val tftypes.Value, target reflect.Value, opts Options, path *tftypes.AttributePath) (reflect.Value, error) {
+	// if this isn't a valid reflect.Value, bail before we accidentally
+	// panic
+	if !target.IsValid() {
+		return target, path.NewErrorf("invalid target")
+	}
+	// if this is an attr.Value, build the type from that
+	if target.Type().Implements(reflect.TypeOf((*attr.Value)(nil)).Elem()) {
```

Comment:
Why not `if _, ok := target.(attr.Value); ok` here?

Reply:
Because target _doesn't_ implement attr.Value, the interface _inside_ it does. I could use `if _, ok := target.Interface().(attr.Value); ok`, but [`.Interface()` can panic](https://golang.org/pkg/reflect/#Value.Interface) so if I don't _have_ to use it, I'd rather stick with the `Implements` method so we don't wind up in a situation in a few months/years where we accidentally access unexported struct fields and panic.

Reply:
(we could guard it with a `.CanInterface()`, but `Implements()` just seemed cleaner, even though the code is uglier.)